data_source.go

```diff
@@ -103,6 +108,97 @@ func (ds *AnySource) StartRun() error {
 	return nil
 }
 
+// makeDirectory creates directory of the form basepath/20060102/000 where
+// the 3-digit subdirectory counts separate file-writing occasions.
+// It also returns the formatting code for use in an Sprintf call
+// basepath/20060102/000/20060102_run000_%s.ljh and an error, if any.
+func makeDirectory(basepath string) (string, error) {
+	today := time.Now().Format("20060102")
+	todayDir := fmt.Sprintf("%s/%s", basepath, today)
+	if err := os.MkdirAll(todayDir, 0755); err != nil {
+		return "", err
+	}
+	for i := 0; i < 10000; i++ {
+		thisDir := fmt.Sprintf("%s/%4.4d", todayDir, i)
+		_, err := os.Stat(thisDir)
+		if os.IsNotExist(err) {
+			if err2 := os.MkdirAll(thisDir, 0755); err2 != nil {
+				return "", err
+			}
+			return fmt.Sprintf("%s/%s_run%4.4d_%%s.ljh", thisDir, today, i), nil
+		}
+	}
+	return "", fmt.Errorf("out of 4-digit ID numbers for today in %s", todayDir)
+}
+
+// WriteControl changes the data writing start/stop/pause/unpause state
+func (ds *AnySource) WriteControl(config *WriteControlConfig) error {
+	if strings.HasPrefix(config.Request, "Pause") {
+		for _, dsp := range ds.processors {
+			dsp.DataPublisher.SetPause(true)
+		}
+		ds.writingState.Paused = true
+
+	} else if strings.HasPrefix(config.Request, "Unpause") {
+		for _, dsp := range ds.processors {
+			dsp.DataPublisher.SetPause(false)
+		}
+		ds.writingState.Paused = false
+
+	} else if strings.HasPrefix(config.Request, "Stop") {
+		for _, dsp := range ds.processors {
+			dsp.DataPublisher.RemoveLJH22()
+			dsp.DataPublisher.RemoveLJH3()
+		}
+		ds.writingState.Active = false
+		ds.writingState.Paused = false
+		ds.writingState.Filename = ""
+
+	} else if strings.HasPrefix(config.Request, "Start") {
+		if config.FileType != "LJH2.2" {
+			return fmt.Errorf("WriteControl FileType=%q, needs to be %q",
+				config.FileType, "LJH2.2")
+		}
+		filenamePattern, err := makeDirectory(config.Path)
```

Comment:
It looks like the client (dastard-commander) is responsible for sending the same path with each file writing request?

Reply:
Yes. What else could we do?

I guess we could use the previous value whenever `config.Path` is the empty string. I'll investigate.